### PR TITLE
Add format getter to DecoratorBlockNode and getSelectionFormat to playground

### PR DESCRIPTION
**Pull request by zignis**
getSelectionFormat allows you to get the common element format of every node that inherits the format style from its parent in the selection. Additionally, the toolbar in the playground has been modified to add an active class to the alignment dropdown item when the selection matches one of the alignment presets. I have also made a few typo fixes. Tests have not been written yet, but I am waiting for your suggestions.

### CODE REVIEW

File: packages/lexical-utils/src/index.ts [hunk 1]
```diff
@@ -511,3 +513,31 @@ export function isHTMLElement(x: Node | EventTarget): x is HTMLElement {
   // @ts-ignore-next-line - strict check on nodeType here should filter out non-Element EventTarget implementors
   return x.nodeType === 1;
 }
+
+/**
+ * @param selection - A range selection.
+ * @returns The common element format type that every node in the range selection inherits from its parent.
+ */
+export function getSelectionFormat(
```
[the remark this answers is not in the document]
Comment (zurfyx) [reply packages/lexical-utils/src/index.ts]: > Should it loop through every node and check for its format?

Based on other editors, this sounds reasonable to me!

[the remark this answers is not in the document]
Comment (zignis) [reply packages/lexical-utils/src/index.ts]: I've added the getSelectionFormat function back to the playground, with some guardrails like ignoring the root. I guess this should do the trick.

[the remark this answers is not in the document]
Comment (zurfyx) [reply packages/lexical-utils/src/index.ts]: Sorry for the back and forth. In reflection, I think your original idea is good, let's roll back to it. I would just add, let's make the code a getter for `FORMAT_ELEMENT_COMMAND`, you will have to tweak some things to make it work with block elements and NodeSelection. Pasting the interesting bits you have to look at:

Rich-text

```
editor.registerCommand<ElementFormatType>(
      FORMAT_ELEMENT_COMMAND,
      (format) => {
        const selection = $getSelection();
        if (!$isRangeSelection(selection) && !$isNodeSelection(selection)) {
          return false;
        }
        const nodes = selection.getNodes();
        for (const node of nodes) {
          const element = $getNearestBlockElementAncestorOrThrow(node);
          element.setFormat(format);
        }
        return true;
      },
      COMMAND_PRIORITY_EDITOR,
    ),
```

LexicalBlockWithAlignableContent

```
        FORMAT_ELEMENT_COMMAND,
        (formatType) => {
          if (isSelected) {
            const selection = $getSelection();

            if ($isNodeSelection(selection)) {
              const node = $getNodeByKey(nodeKey);

              if ($isDecoratorBlockNode(node)) {
                node.setFormat(formatType);
              }
            } else if ($isRangeSelection(selection)) {
              const nodes = selection.getNodes();

              for (const node of nodes) {
                if ($isDecoratorBlockNode(node)) {
                  node.setFormat(formatType);
                } else {
                  const element = $getNearestBlockElementAncestorOrThrow(node);
                  element.setFormat(formatType);
                }
              }
            }

            return true;
          }

          return false;
        },
        COMMAND_PRIORITY_LOW,
      ),
```

I would also rename the method to `$getBlockFormat` based on the other existing block functions (i.e. `$setBlocksType`)

**Comment by M55-ai** [comment]
> getSelectionFormat allows you to get the common element format of every node that inherits the format style from its parent in the selection. Additionally, the toolbar in the playground has been modified to add an active class to the alignment dropdown item when the selection matches one of the alignment presets. I have also made a few typo fixes. Tests have not been written yet, but I am waiting for your suggestions.

File: packages/lexical-utils/src/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (zignis) [reply packages/lexical-utils/src/index.ts]: Added the `$getBlocksFormat` method to `@lexical/selection`. I also had to add a getter for element format type to `DecoratorBlockNode` in `@lexical/react`.

File: packages/lexical-react/src/LexicalDecoratorBlockNode.ts [hunk 2]
```diff
@@ -52,6 +52,11 @@ export class DecoratorBlockNode extends DecoratorNode<JSX.Element> {
     self.__format = format;
   }
 
+  getFormatType(): ElementFormatType {
```
Comment (zurfyx) [inline packages/lexical-react/src/LexicalDecoratorBlockNode.ts:55]:
```suggestion
  getFormat(): ElementFormatType {
```

**Review by zurfyx — APPROVED** [review]
LGTM, can you look into the failing E2E tests please?

**Comment by zignis** [comment]
> LGTM, thank you for iterating on this, can you look into the failing E2E tests please?

Tried looking into them, they seem to depend on other mods and would require a thorough review.

**Review by zurfyx — COMMENTED** [review]
It turns out that `getNearestBlockElementAncestorOrThrow` was causing this to crash on a non-block DecoratorNode as it's already as the highest level. Arguably, it should be a BlockDecoratorNode in the first place but we can handle this gracefully. 

There's still another issue with the Toolbar code, try playing with horizontal rule, deleting backward and you will notice it does not delete. Could you debug the code you introduce in the Toolbar to see if you can find what's causing this?

You can run the failing E2Es locally like this `npm run test-e2e-chromium -- HorizontalRule.spec`

File: packages/lexical-playground/src/utils/getSelectionFormat.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  const formatSet = new Set<ElementFormatType>();
+
+  for (const node of nodes) {
+    if ($isDecoratorBlockNode(node)) {
+      formatSet.add(node.getFormat());
+    } else {
+      const element = $getNearestBlockElementAncestorOrThrow(node);
+      formatSet.add(element.getFormatType());
+    }
+  }
```
Comment (zurfyx) [inline packages/lexical-playground/src/utils/getSelectionFormat.ts:30]:
```suggestion
  for (const node of nodes) {
    const topNode = node.getTopLevelElementOrThrow();
    if ($isDecoratorBlockNode(topNode)) {
      formatSet.add(node.getFormat());
    } else if ($isElementNode(topNode)) {
      formatSet.add(topNode.getFormatType());
    }
  }
```

**Comment by zignis** [comment]
> It turns out that `getNearestBlockElementAncestorOrThrow` was causing this to crash on a non-block DecoratorNode as it's already as the highest level. Arguably, it should be a BlockDecoratorNode in the first place but we can handle this gracefully.
> 
> There's still another issue with the Toolbar code, try playing with horizontal rule, deleting backward and you will notice it does not delete. Could you debug the code you introduce in the Toolbar to see if you can find what's causing this?
> 
> You can run the failing E2Es locally like this `npm run test-e2e-chromium -- HorizontalRule.spec`

I did debug it locally (with the --debug arg) earlier and I turns out it was throwing due to `getNearestBlockElementAncestorOrThrow`, sorry as I haven't dug deep enough into lexical's core yet, but this does seem to fix some of the e2e tests. I'll try looking into others.

**Comment by zignis** [comment]
Hi @zurfyx , can we move forward with this? I can see the functionality has been implemented in the toolbar: https://github.com/facebook/lexical/blob/fa68893288693cb9be0b0b5d868b3f02fee14b3b/packages/lexical-playground/src/plugins/ToolbarPlugin/index.tsx#L596C1-L600C9

```tsx
      setElementFormat(
        ($isElementNode(node)
          ? node.getFormatType()
          : parent?.getFormatType()) || 'left',
      );
```       

The issue here (I might be wrong) is that it still does not account for a selection of nodes with different formats, and it falls back to `left`.

**Comment by potatowagon** [comment]
Closing this PR due to staleness! If there are new updates, please reopen the PR.

**Pull request closed without merging**